proposals/0048-container-api.md

```diff
@@ -0,0 +1,221 @@
+- Possibly unblock `.mdx` compiledContent/html output?
+- Support Astro framework renderers (`@astrojs/*`) if possible
+
+# Non-Goals
+
+- Provide a way to **import** `.astro` components. Users will be responsible to provide a compiled component.
+- Provide a way to **import** `astro.config.(mjs|mts)` out of the box.
+
+# Detailed Design
+
+## Preface
+
+**This** RFC will have a smaller scoped compared to what users have envisioned. The reason why the scope shrunk is that 
+I want to ship a smaller feature on the surface, which will allow us, later to enhance it based on user's feedback and use cases.
+
+That's why the first iteration of the APIs won't provide a built-in way to compile Astro components.
+
+## Internal concepts
+
+The API took a lot of time to land because it required some internal refactors. Thanks to these refactors, we will be able
+to land this API by using the **very same engine** that Astro uses under the hood.
+
+Eventually, we landed to a common and abstract concept called **Pipeline**. A pipeline is simply an abstract
+class that is responsible to render any kind of Astro route: page, endpoint, redirect, virtual.
+
+Each pipeline inside the Astro codebase (dev, build, SSR and now test) is responsible to collect the important information in different way,
+but eventually each route must be rendered using the same data, which are:
+- **Manifest**: this is a serializable/deserializable version of our "configuration", however it also contains more information, such as renderers, client directives, styles.
+- **Request**: a standard https://developer.mozilla.org/en-US/docs/Web/API/Request object.
+- **RouteData**: a type that identifies a generic Astro route. It contains information such as the type of route, the associated component, etc.
+- **ComponentInstance**: The instance of an Astro component. This is the **compiled** component. Internally, an Astro component undergoes various transformations via Vite, and it eventually gets consumed by our rendering engine (if it's a page).
+
+## Create a container
+
+A container is class exposed via the `astro/container` specifier. Users **shouldn't** create a container using the `new` instance, but they should use he static function `create`:
+
+```js
+const container = await AstroContainer.create()
+```
+
+The reason why `create` returns a promise is because internally we will validate and parse the user configuration. As a developer, you will be able to pass the same configuration that Astro uses:
+
+```js
+import astroConfig from "../src/astro.config.mjs";
+
+const container = await AstroContainer.create({
+    astroConfig
+})
+```
+
+The function `create` will throw an error if there's a validation error.
+
+> [!NOTE]
+> If you use a TypeScript file for your configuration, you are responsible for loading and transforming it to JavaScript.
+
+
+The `container` binding will expose a `renderToString` that accepts a Astro component and will return a `Response` object:
```

Comment:
Should it be called `renderToString` if it doesn't.. return a string? Maybe it'd be better to have a `renderToResponse` or something, and a `renderToString` wrapper that does the response.body dance for you.

Reply:
My bad. It returns a `string` :P

Reply:
+1, agreed, `renderToString` is a common need for testing, so makes sense to have a helper for this case.